### PR TITLE
Hotfix: Upgrade go-autorest to v9.5.2

### PR DESCRIPTION
**Pull request by metacpp**

### CODE REVIEW

**Comment by sebastus** [comment]
the part that I know about (sender.go) looks good.

**Review by tombuildsstuff — APPROVED** [review]
LGTM - just running the tests now 👍

**Comment by tombuildsstuff** [comment]
Looks like there's a bug in `go-autorest v5.4.1`:

```
Test ended in panic.

------- Stdout: -------
=== RUN   TestAccAzureRMSqlElasticPool_basic

------- Stderr: -------
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x702133]

goroutine 414 [running]:
github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/Azure/go-autorest/autorest.DoRetryForStatusCodes.func1.1(0xc4204b8000, 0xc4205e7300, 0x0, 0xc420054d08)
    /opt/teamcity-agent/work/b5bd5badde638102/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/Azure/go-autorest/autorest/sender.go:235 +0x1a3
github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/Azure/go-autorest/autorest.SenderFunc.Do(0xc4205e7300, 0xc4204b8000, 0xc420054d08, 0x1, 0x1)
    /opt/teamcity-agent/work/b5bd5badde638102/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/Azure/go-autorest/autorest/sender.go:38 +0x30
github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/Azure/go-autorest/autorest.SendWithSender(0x2243780, 0xc4203577c0, 0xc4204b8000, 0xc420054d08, 0x1, 0x1, 0x1280000000000, 0x1, 0xa)
    /opt/teamcity-agent/work/b5bd5badde638102/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/Azure/go-autorest/autorest/sender.go:80 +0x75
github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/Azure/go-autorest/autorest/azure.DoRetryWithRegistration.func1.1(0xc4204b8000, 0xc4203a4000, 0x0, 0xc420054f70)
    /opt/teamcity-agent/work/b5bd5badde638102/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/Azure/go-autorest/autorest/azure/rp.go:40 +0x1d4
github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/Azure/go-autorest/autorest.SenderFunc.Do(0xc4203a4000, 0xc4204b8000, 0xc420054f70, 0x1, 0x1)
    /opt/teamcity-agent/work/b5bd5badde638102/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/Azure/go-autorest/autorest/sender.go:38 +0x30
github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/Azure/go-autorest/autorest.SendWithSender(0x2243780, 0xc4203577c0, 0xc4204b8000, 0xc420054f70, 0x1, 0x1, 0xdf8475800, 0xd18c2e2800, 0x3)
    /opt/teamcity-agent/work/b5bd5badde638102/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/Azure/go-autorest/autorest/sender.go:80 +0x75
github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/Azure/azure-sdk-for-go/arm/resources/resources.ProvidersClient.ListSender(0x223e640, 0xc420346e60, 0x2243ec0, 0xc4202d2660, 0x0, 0x0, 0xdf8475800, 0xd18c2e2800, 0x3, 0x6fc23ac00, ...)
    /opt/teamcity-agent/work/b5bd5badde638102/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/Azure/azure-sdk-for-go/arm/resources/resources/providers.go:166 +0x110
github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/Azure/azure-sdk-for-go/arm/resources/resources.ProvidersClient.List(0x223e640, 0xc420346e60, 0x2243ec0, 0xc4202d2660, 0x0, 0x0, 0xdf8475800, 0xd18c2e2800, 0x3, 0x6fc23ac00, ...)
    /opt/teamcity-agent/work/b5bd5badde638102/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/Azure/azure-sdk-for-go/arm/resources/resources/providers.go:123 +0x105
github.com/terraform-providers/terraform-provider-azurerm/azurerm.providerConfigure.func1(0xc42010c150, 0xc42010c150, 0xc42018c6c0, 0xc4202d2420, 0x0)
    /opt/teamcity-agent/work/b5bd5badde638102/src/github.com/terraform-providers/terraform-provider-azurerm/azurerm/provider.go:226 +0x577
github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform/helper/schema.(*Provider).Configure(0xc42010de30, 0xc42018c6c0, 0x7, 0xc42018c6c0)
    /opt/teamcity-agent/work/b5bd5badde638102/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform/helper/schema/provider.go:240 +0xdf
github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform/terraform.(*BuiltinEvalContext).ConfigureProvider(0xc420368000, 0xc420018fa1, 0x7, 0xc42018c6c0, 0x0, 0x31)
    /opt/teamcity-agent/work/b5bd5badde638102/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform/terraform/eval_context_builtin.go:157 +0xe8
github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform/terraform.(*EvalConfigProvider).Eval(0xc4202d2280, 0x224d620, 0xc420368000, 0x2, 0x2, 0x1778c6d, 0x4)
    /opt/teamcity-agent/work/b5bd5badde638102/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform/terraform/eval_provider.go:65 +0x53
github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform/terraform.EvalRaw(0x223fbc0, 0xc4202d2280, 0x224d620, 0xc420368000, 0xc42006c0f0, 0x2, 0xc4205c8210, 0x2b)
    /opt/teamcity-agent/work/b5bd5badde638102/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform/terraform/eval.go:53 +0x173
github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform/terraform.(*EvalSequence).Eval(0xc4202d22a0, 0x224d620, 0xc420368000, 0x2, 0x2, 0x1778c6d, 0x4)
    /opt/teamcity-agent/work/b5bd5badde638102/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform/terraform/eval_sequence.go:14 +0x7e
github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform/terraform.EvalRaw(0x2240340, 0xc4202d22a0, 0x224d620, 0xc420368000, 0x2b, 0x0, 0x0, 0x2b)
    /opt/teamcity-agent/work/b5bd5badde638102/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform/terraform/eval.go:53 +0x173
github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform/terraform.(*EvalOpFilter).Eval(0xc42018c420, 0x224d620, 0xc420368000, 0x2, 0x2, 0x1778c6d, 0x4)
    /opt/teamcity-agent/work/b5bd5badde638102/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform/terraform/eval_filter_operation.go:37 +0x4c
github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform/terraform.EvalRaw(0x2240100, 0xc42018c420, 0x224d620, 0xc420368000, 0x0, 0x0, 0x0, 0x0)
    /opt/teamcity-agent/work/b5bd5badde638102/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform/terraform/eval.go:53 +0x173
github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform/terraform.(*EvalSequence).Eval(0xc4202d22c0, 0x224d620, 0xc420368000, 0x2, 0x2, 0x1778c6d, 0x4)
    /opt/teamcity-agent/work/b5bd5badde638102/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform/terraform/eval_sequence.go:14 +0x7e
github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform/terraform.EvalRaw(0x2240340, 0xc4202d22c0, 0x224d620, 0xc420368000, 0x15c68e0, 0x224e503, 0x1574b00, 0xc420346180)
    /opt/teamcity-agent/work/b5bd5badde638102/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform/terraform/eval.go:53 +0x173
github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform/terraform.Eval(0x2240340, 0xc4202d22c0, 0x224d620, 0xc420368000, 0xc4202d22c0, 0x2240340, 0xc4202d22c0, 0x48e954)
    /opt/teamcity-agent/work/b5bd5badde638102/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform/terraform/eval.go:34 +0x4d
github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform/terraform.(*Graph).walk.func1(0x16b68a0, 0xc42000f7b8, 0x0, 0x0)
    /opt/teamcity-agent/work/b5bd5badde638102/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform/terraform/graph.go:126 +0xd0d
github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform/dag.(*Walker).walkVertex(0xc4203cafc0, 0x16b68a0, 0xc42000f7b8, 0xc420694200)
    /opt/teamcity-agent/work/b5bd5badde638102/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform/dag/walk.go:387 +0x3c1
created by github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform/dag.(*Walker).Update
    /opt/teamcity-agent/work/b5bd5badde638102/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform/dag/walk.go:310 +0x1364
```

File: vendor/github.com/Azure/go-autorest/autorest/sender.go [hunk 1]
```diff
@@ -215,20 +215,26 @@ func DoRetryForStatusCodes(attempts int, backoff time.Duration, codes ...int) Se
...
+				if err == nil && !ResponseHasStatusCode(resp, codes...) || IsTokenRefreshError(err) {
 					return resp, err
 				}
 				delayed := DelayWithRetryAfter(resp, r.Cancel)
 				if !delayed {
 					DelayForBackoff(backoff, attempt, r.Cancel)
 				}
+				// don't count a 429 against the number of attempts
+				// so that we continue to retry until it succeeds
+				if resp.StatusCode != http.StatusTooManyRequests {
```
Comment (tombuildsstuff) [inline vendor/github.com/Azure/go-autorest/autorest/sender.go]: 👋🏻 @jhendrixMSFT @marstr we're seeing a crash here - we handle checking for this internally to see if there was a dropped connection from Azure (which happens often enough that we need to handle it) - any chance that we could check if `resp != nil` before checking the status code?

**Review by tombuildsstuff — DISMISSED** [review]
The acceptance tests have showed that this panics when the connection is reset/dropped (whereas the current version of Go-AutoRest will instead return an error, which can be handled gracefully).

Can we update this to check that `resp` isn't nil before accessing it, so that we can handle this in a more graceful fashion?

File: vendor/github.com/Azure/go-autorest/autorest/sender.go (hunk 1 above)
Comment (jhendrixMSFT) [reply vendor/github.com/Azure/go-autorest/autorest/sender.go]: I will get this fixed and release a v9.5.2 today.

Comment (tombuildsstuff) [reply vendor/github.com/Azure/go-autorest/autorest/sender.go]: awesome - thanks @jhendrixMSFT :)

Comment (jhendrixMSFT) [reply vendor/github.com/Azure/go-autorest/autorest/sender.go]: The fix is in an the release has been tagged.  https://github.com/Azure/go-autorest/releases/tag/v9.5.2

**Comment by mbfrahry** [comment]
The above test passes now. 

```
make testacc TEST=./azurerm TESTARGS="-run=TestAccAzureRMSqlElasticPool_basic"
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./azurerm -v -run=TestAccAzureRMSqlElasticPool_basic -timeout 180m
=== RUN   TestAccAzureRMSqlElasticPool_basic
--- PASS: TestAccAzureRMSqlElasticPool_basic (134.18s)
PASS
ok  	github.com/terraform-providers/terraform-provider-azurerm/azurerm	134.206s
```

**Review by mbfrahry — APPROVED** [review]
LGTM

File: CHANGELOG.md [hunk 2]
```diff
@@ -22,6 +22,12 @@ IMPROVEMENTS:
 * `azurerm_virtual_machine_scale_set` - Support for updating the customData field [GH-559]
 * `azurerm_virtual_machine_scale_set` - ensuring changes to the `settings` object are detected [GH-609]
 
+## 0.3.4 (December 12, 2017)
+
+IMPROVEMENTS:
+
+* Too Many Request: Upgrade go-autorest to v9.5.2 to keep retrying all the time while meeting 429 status code. ([#617](https://github.com/terraform-providers/terraform-provider-azurerm/pull/617))
+
```
Comment (tombuildsstuff) [inline CHANGELOG.md]: @metacpp heads up that we don't manage this file through PR's (also there is no v0.3.4 of the Provider) - so I'm going to push a commit to revert this file/change (and update it as we normally would)

**Review by tombuildsstuff — APPROVED** [review]
LGTM - thanks for re-vendoring this 👍

**Comment by tombuildsstuff** [comment]
Meant to add - ran the test suite and this seems fine with the changes in AutoRest v9.5.2 👍

**Pull request merged**